Mathlib/Analysis/SpecialFunctions/Trigonometric/Chebyshev/RootsExtrema.lean

```diff
@@ -0,0 +1,328 @@
+  simp
+
+theorem one_lt_negOnePow_mul_eval_T_real {n : ℤ} (hn : n ≠ 0) {x : ℝ} (hx : x < -1) :
+    1 < (-1) ^ n * (T ℝ n).eval x := by
+  rw [← neg_neg x, T_eval_neg]
+  convert one_lt_eval_T_real hn (lt_neg_of_lt_neg hx)
+  rw [Int.cast_negOnePow, ← mul_assoc, ← mul_zpow]
+  simp
+
+theorem one_le_abs_eval_T_real (n : ℤ) {x : ℝ} (hx : 1 ≤ |x|) :
+    1 ≤ |(T ℝ n).eval x| := by
+  cases le_abs.mp hx with
+  | inl hx => exact one_le_eval_T_real n hx |> .inl |> le_abs.mpr
+  | inr hx =>
+    have := one_le_negOnePow_mul_eval_T_real n (le_neg_of_le_neg hx)
+    have : 1 ≤ |(-1) ^ n * (T ℝ n).eval x| := by grind
+    convert this using 1
+    simp [abs_mul, abs_zpow]
+
+theorem one_lt_abs_eval_T_real {n : ℤ} (hn : n ≠ 0) {x : ℝ} (hx : 1 < |x|) :
+    1 < |(T ℝ n).eval x| := by
+  cases lt_abs.mp hx with
+  | inl hx => exact one_lt_eval_T_real hn hx |> .inl |> lt_abs.mpr
+  | inr hx =>
+    have := one_lt_negOnePow_mul_eval_T_real hn (lt_neg_of_lt_neg hx)
+    have : 1 < |(-1) ^ n * (T ℝ n).eval x| := by grind
+    convert this using 1
+    simp [abs_mul, abs_zpow]
+
+theorem abs_eval_T_real_le_one_iff {n : ℤ} (hn : n ≠ 0) (x : ℝ) :
+    |x| ≤ 1 ↔ |(T ℝ n).eval x| ≤ 1 := by
+  constructor
+  · intro hx; exact abs_eval_T_real_le_one n hx
+  · intro hx; contrapose! hx; exact one_lt_abs_eval_T_real hn hx
+
+theorem abs_eval_T_real_eq_one_iff {n : ℕ} (hn : n ≠ 0) (x : ℝ) :
+    |(T ℝ n).eval x| = 1 ↔ ∃ k ≤ n, x = cos (k * π / n) := by
+  constructor
+  · intro hTx
+    have hx := (abs_eval_T_real_le_one_iff (Nat.cast_ne_zero.mpr hn) x).mpr (le_of_eq hTx)
+    rw [← cos_arccos (neg_le_of_abs_le hx) (le_of_max_le_left hx), T_real_cos,
+      Int.cast_natCast] at hTx
+    have : ∃ (k : ℤ), k * π = n * arccos x := by
+      cases (abs_eq zero_le_one).mp hTx
+      case inl h =>
+        obtain ⟨k, hk⟩ := (cos_eq_one_iff _).mp h
+        use 2 * k
+        grind
+      case inr h =>
+        obtain ⟨k, hk⟩ := cos_eq_neg_one_iff.mp h
+        use 2 * k + 1
+        grind
+    obtain ⟨k, hk⟩ := this
+    replace hk : arccos x = (k * π) / n := by aesop
+    have k_nonneg : 0 ≤ k := by
+      suffices 0 ≤ (k : ℝ) by norm_cast at this
+      have : 0 ≤ k * π / n := by grind [arccos_nonneg]
+      linear_combination (norm := field_simp) (n / π) * this
+      grind
+    have k_le_n : k ≤ n := by
+      suffices (k : ℝ) ≤ n by norm_cast at this
+      have : k * π / n ≤ π := by grind [arccos_le_pi]
+      linear_combination (norm := field_simp) (n / π) * this
+      grind
+    use k.toNat
+    refine ⟨Int.toNat_le.mpr k_le_n, ?_⟩
+    rw [← cos_arccos (neg_le_of_abs_le hx) (le_of_max_le_left hx), hk]
+    congr
+    exact (Int.toNat_of_nonneg k_nonneg).symm
+  · rintro ⟨k, hk, hx⟩
+    have : ((n : ℤ) : ℝ) * (k * π / n) = (k : ℤ) * π := by norm_cast; field_simp
+    rw [hx, T_real_cos, this, cos_int_mul_pi, abs_neg_one_zpow]
+
+theorem eval_T_real_eq_one_iff {n : ℕ} (hn : n ≠ 0) (x : ℝ) :
+    (T ℝ n).eval x = 1 ↔ ∃ k ≤ n, Even k ∧ x = cos (k * π / n) := by
```

Comment:
You should be able to simplify this lemma and the next one a lot with this lemma:
```lean
theorem eval_T_real_cos_int_mul_pi_div {k : ℤ} {n : ℕ} (hn : n ≠ 0) :
    (T ℝ n).eval (cos (k * π / n)) = k.negOnePow := by
  rw [T_real_cos, Int.cast_negOnePow]
  convert Real.cos_int_mul_pi k using 2 
  simp [field]
```